runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java

```diff
@@ -123,4 +130,25 @@ public String create(PipelineOptions options) {
       return normalizedAppName + "-" + normalizedUserName + "-" + datePart;
     }
   }
+
+  /**
+   * Returns a default staging location under {@link GcpOptions#getGcpTempLocation}.
+   */
+  public static class StagingLocationFactory implements DefaultValueFactory<String> {
+
+    @Override
+    public String create(PipelineOptions options) {
+      String gcpTempLocation = options.as(GcpOptions.class).getGcpTempLocation();
+      checkArgument(!isNullOrEmpty(gcpTempLocation),
+          "Missing required value: at least one of tempLocation, gcpTempLocation "
+          + "or stagingLocation must be set.");
```

Comment:
can be misleading, because `tempLocation` can be set to `file://blah` and this will still not work.

I would just say `Error constructing default value for stagingLocation: gcpTempLocation is missing. Either stagingLocation must be set explicitly or a valid value must be provided for gcpTempLocation`)